zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/distribution/CommandDistributionBehavior.java

```diff
@@ -67,49 +103,57 @@ public <T extends UnifiedRecordValue> void distributeCommand(
     stateWriter.appendFollowUpEvent(
         distributionKey, CommandDistributionIntent.STARTED, distributionRecord);
 
-    otherPartitions.forEach(
-        (partition) ->
-            distributeToPartition(command, partition, distributionRecord, distributionKey));
+    partitions.forEach(
+        (partition) -> distributeToPartition(partition, distributionRecord, distributionKey));
   }
 
   private <T extends UnifiedRecordValue> void distributeToPartition(
-      final TypedRecord<T> command,
       final int partition,
       final CommandDistributionRecord distributionRecord,
       final long distributionKey) {
     final var valueType = distributionRecord.getValueType();
-    final var commandValue = distributionRecord.getCommandValue();
+    final var intent = distributionRecord.getIntent();
+
     // We don't need the actual record in the DISTRIBUTING event applier. In order to prevent
     // reaching the max message size we don't set the record value here.
     stateWriter.appendFollowUpEvent(
         distributionKey,
         CommandDistributionIntent.DISTRIBUTING,
-        new CommandDistributionRecord()
+        commandDistributionDistributing
             .setPartitionId(partition)
             .setValueType(valueType)
-            .setIntent(command.getIntent()));
+            .setIntent(intent));
+
+    // This getter makes a hard copy of the command value, which we need to send the command to the
+    // other partition in a side effect. It does not appear to be possible to reuse a single
+    // instance for distributing to all partitions in the form of a method parameter, but it's not
+    // fully clear why that leads to problems. We suspect that it's because the command value is a
+    // mutable object, and it is somehow modified before the side effect is executed. Instead, we
+    // have to copy this value for every partition.
+    final var commandValue = distributionRecord.getCommandValue();
 
     sideEffectWriter.appendSideEffect(
         () -> {
           interPartitionCommandSender.sendCommand(
-              partition,
-              command.getValueType(),
-              command.getIntent(),
-              distributionKey,
-              commandValue);
+              partition, valueType, intent, distributionKey, commandValue);
           return true;
         });
   }
 
-  public <T extends UnifiedRecordValue> void acknowledgeCommand(
-      final long distributionKey, final TypedRecord<T> command) {
+  /**
+   * Acknowledges that a command was distributed to another partition successfully.
+   *
+   * @param command the command that was distributed
+   */
+  public <T extends UnifiedRecordValue> void acknowledgeCommand(final TypedRecord<T> command) {
+    final long distributionKey = command.getKey();
     final var distributionRecord =
```

Comment:
❓Why we didn't create a re-usable record for the acknowledged command as we did refactoring for others?